tests/operators/test_python.py

```diff
@@ -311,6 +312,314 @@ def func(**context):
+        recorded_calls = []
+
+        task = task_decorator(
+            # a Mock instance cannot be used as a callable function or test fails with a
+            # TypeError: Object of type Mock is not JSON serializable
+            build_recording_function(recorded_calls),
+            dag=self.dag
+        )
+        task(an_int=4, a_date=date(2019, 1, 1), a_templated_string="dag {{dag.dag_id}} ran on {{ds}}.")
+        self.dag.create_dagrun(
+            run_id='manual__' + DEFAULT_DATE.isoformat(),
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        self.assertEqual(1, len(recorded_calls))
+        self._assert_calls_equal(
+            recorded_calls[0],
+            Call(an_int=4,
+                 a_date=date(2019, 1, 1),
+                 a_templated_string="dag {} ran on {}.".format(
+                     self.dag.dag_id, DEFAULT_DATE.date().isoformat()))
+        )
+
+    def test_copy_in_dag(self):
+        with self.dag:
+            self.do_run()
+            self.assertListEqual(['do_run'], self.dag.task_ids)
+            do_run_1 = self.do_run.copy()
+            do_run_2 = self.do_run.copy()
+        self.assertEqual(do_run_1.task_id, 'do_run__1')
+        self.assertEqual(do_run_2.task_id, 'do_run__2')
+
+    def test_copy(self):
+        do_run_1 = self.do_run.copy()
+        do_run_2 = self.do_run.copy()
+        with self.dag:
+            self.do_run()
+            self.assertListEqual(['do_run'], self.dag.task_ids)
+            do_run_1()
+            do_run_2()
+
+        self.assertEqual('do_run__1', do_run_1.task_id)
+        self.assertEqual('do_run__2', do_run_2.task_id)
+
+    def test_dict_outputs(self):
+        @task_decorator(multiple_outputs=True)
+        def return_dict(number: int):
+            return {
+                'number': number + 1,
+                43: 43
+            }
+
+        test_number = 10
+        with self.dag:
+            return_dict(test_number)
+
+        dr = self.dag.create_dagrun(
+            run_id="manual__",
+            start_date=timezone.utcnow(),
+            execution_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+
+        return_dict.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        tis = dr.get_task_instances()
+        for ti in tis:
+            if ti.task_id == 'return_dict':
+                self.assertEqual(
+                    test_number + 1,
+                    ti.xcom_pull(key='number')
+                )
+                self.assertEqual(
+                    43,
+                    ti.xcom_pull(key='43')
+                )
+                self.assertDictEqual(
+                    {
+                        'number': test_number + 1,
+                        '43': 43
+                    },
+                    ti.xcom_pull()
+                )
```

Comment:
Can we instead of iterating over tis just get the ti that interests us? Also, will there be many tis?